custom_methods.rb

```diff
@@ -0,0 +1,148 @@
+# rubocop:disable Style/CaseEquality
+# rubocop:disable Metrics/ModuleLength
+module Enumerable
+  def change_self
+    return Array self if self.class == Range
+
+    self
+  end
+
+  def my_each
+    my_array = change_self
+    return to_enum unless block_given?
+
+    i = 0
+    while i < my_array.length
+      yield my_array[i]
+      i += 1
+    end
+    my_array
+  end
+
+  def my_each_with_index
+    my_array = change_self
+    return to_enum unless block_given?
+
+    i = 0
+    while i < my_array.length
+      yield my_array[i], i
+      i += 1
+    end
+    my_array
+  end
+
+  def my_select
+    my_array = change_self
+    return to_enum unless block_given?
+
+    arr = []
+    my_array.my_each { |x| arr.push(x) if yield x }
+    arr
+  end
+
+  def my_all?(word = nil)
+    my_array = change_self
+    if block_given?
+      my_array.my_each { |item| return false unless yield item }
+      true
+    elsif word
+      my_array.my_each { |item| return false unless n === item }
+      true
+    else
+      my_all? { |item| item }
+    end
+  end
+
+  def my_any?(regex = false)
+    my_array = change_self
+    if block_given?
+      my_array.my_each { |item| return true if yield item }
+      false
+    elsif regex
+      my_array.my_each { return true if regex === my_array[i] }
+      false
+    else
+      my_any? { |item| item }
+    end
+  end
+
+  def my_none?(regex = false)
+    my_array = change_self
+    if block_given?
+      my_array.my_each { |item| return false if yield item }
+      true
+    elsif regex
+      my_array.my_each { |item| return false if regex === item }
+      true
+    else
+      my_none? { |item| item }
+    end
+  end
+
+  def my_count(item = false)
+    my_array = change_self
+    counter = 0
+    if block_given?
+      my_array.my_each { |elem| counter += 1 if yield elem }
+      counter
+    elsif item
+      my_array.my_each { |elem| counter += 1 if elem == item }
+      counter
+    else
+      my_array.length
+    end
+  end
+
+  def my_map(proc = false)
+    my_array = change_self
+    arr = []
+    if block_given?
+      my_array.my_each { |item| arr.push(yield item) }
+      arr
+    elsif proc
+      my_array.my_each { |item| arr.push(proc.call(item)) }
+      arr
+    else
+      to_enum
+    end
+  end
+
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/PerceivedComplexity
+
+  def my_inject(memo = false, symbol = false)
```

Comment:
* [ ]There should be a working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
compare the answers below.
This is supposed to return an error, compare answers between the two examples
``` (5..10).my_inject(nil) { |product, n| product * n }``` != ```(5..10).inject(nil) { |product, n| product * n }```

Reply:
Hello Forison, first of all, thanks for the code review, I have done the changes for the first two functions, but with this one, I would like to clarify when you say "This is supposed to return an error", the error should be thrown by the code you send? `(5..10).my_inject(nil) { |product, n| product * n } != (5..10).inject(nil) { |product, n| product * n }` or just returned using the function `(5..10).inject(nil) { |product, n| product * n }` ? Any answer or clarification would be greatly appreciated.